shell/platform/android/io/flutter/plugin/editing/ListenableEditingState.java

```diff
@@ -0,0 +1,234 @@
+///
+/// Changing the editing state in a didChangeEditingState callback may cause unexpected behavior.
+//
+// Currently this class does not notify its listeners on spans-only changes (e.g.,
+// Selection.setSelection). Wrap them in a batch edit to trigger a change notification.
+class ListenableEditingState extends SpannableStringBuilder {
+  interface EditingStateWatcher {
+    // Changing the editing state in a didChangeEditingState callback may cause unexpected
+    // behavior.
+    void didChangeEditingState(
+        boolean textChanged, boolean selectionChanged, boolean composingRegionChanged);
+  }
+
+  private static final String TAG = "flutter";
+
+  private int mBatchEditNestDepth = 0;
+  private int mChangeNotificationDepth = 0;
+  private ArrayList<EditingStateWatcher> mListeners = new ArrayList<>();
+  private ArrayList<EditingStateWatcher> mPendingListeners = new ArrayList<>();
+
+  private String mToStringCache;
+
+  private String mTextWhenBeginBatchEdit;
+  private int mSelectionStartWhenBeginBatchEdit;
+  private int mSelectionEndWhenBeginBatchEdit;
+  private int mComposingStartWhenBeginBatchEdit;
+  private int mComposingEndWhenBeginBatchEdit;
+
+  private BaseInputConnection mDummyConnection;
+
+  // The View is only use for creating a dummy BaseInputConnection which is only used in
+  // setComposingRegion. The View needs to have a non-null Context.
+  public ListenableEditingState(TextInputChannel.TextEditState configuration, View view) {
+    super();
+    if (configuration != null) {
+      setEditingState(configuration);
+    }
+
+    Editable self = this;
+    mDummyConnection =
+        new BaseInputConnection(view, true) {
+          @Override
+          public Editable getEditable() {
+            return self;
+          }
+        };
+  }
+
+  /// Starts a new batch edit during which change notifications will be put on hold until all batch
+  /// edits end.
+  ///
+  /// Batch ends nest.
+  public void beginBatchEdit() {
+    mBatchEditNestDepth++;
+    if (mChangeNotificationDepth > 0) {
+      Log.e(TAG, "editing state should not be changed in a listener callback");
+    }
+    if (mBatchEditNestDepth == 1 && !mListeners.isEmpty()) {
+      mTextWhenBeginBatchEdit = toString();
+      mSelectionStartWhenBeginBatchEdit = getSelecionStart();
+      mSelectionEndWhenBeginBatchEdit = getSelecionEnd();
+      mComposingStartWhenBeginBatchEdit = getComposingStart();
+      mComposingEndWhenBeginBatchEdit = getComposingEnd();
+    }
+  }
+
+  /// Ends the current batch edit and flush pending change notifications if the current batch edit
+  /// is not nested (i.e. it was the last ongoing batch edit).
+  public void endBatchEdit() {
+    if (mBatchEditNestDepth == 0) {
+      return;
+    }
+
+    if (mBatchEditNestDepth == 1) {
+      for (final EditingStateWatcher listener : mPendingListeners) {
+        notifyListener(listener, true, true, true);
```

Comment:
Yeah it's in the class documentation.

Reply:
It would be helpful for readers to just include /** textChanged */ comments next to each `true` as it is hard to tell without scrolling and loosing reading context.